### PR TITLE
fix the type of ShallowRenderer.render's return

### PR DESCRIPTION
**Pull request by Tan90Qian**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
v15:<https://github.com/facebook/react/blob/v15.6.2/src/renderers/testing/ReactShallowRenderer.js>; v16:<https://github.com/facebook/react/blob/master/packages/react-test-renderer/src/ReactShallowRenderer.js>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

```javascript
// source code for v16
render(element: ReactElement | null, context: null | Object = emptyObject) {
  ...
  return this.getRenderOutput();
}
```

### CODE REVIEW

**Comment by jgoz** [comment]
I'm slightly hesitant on this change because the official API documentation treats it as `void` (though doesn't say that explicitly) and guides you to use `getRenderOutput()`. [This line](https://github.com/facebook/react/blob/master/packages/react-test-renderer/src/ReactShallowRenderer.js#L525-L527) also gives me pause.

I guess this is a question of whether the typings should more closely reflect the public API or the actual implementation.

**Comment by Tan90Qian** [comment]
```
Reference
shallowRenderer.render()
You can think of the shallowRenderer as a “place” to render the component you’re testing, and from which you can extract the component’s output.

shallowRenderer.render() is similar to ReactDOM.render() but it doesn’t require DOM and only renders a single level deep. This means you can test components isolated from how their children are implemented.
```
<https://reactjs.org/docs/shallow-renderer.html#shallowrendererrender>
the document on reactjs.org don't show what the api returns, and it don't have an example.

```javascript
it('should have shallow rendering', () => {
    class SomeComponent extends React.Component {
      render() {
        return (
          <div>
            <span className="child1" />
            <span className="child2" />
          </div>
        );
      }
    }

    const shallowRenderer = createRenderer();
    const result = shallowRenderer.render(<SomeComponent />);

    expect(result.type).toBe('div');
    expect(result.props.children).toEqual([
      <span className="child1" />,
      <span className="child2" />,
    ]);
  });
```
<https://github.com/facebook/react/blob/master/packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js>
Here is test case which uses the result to do something later.

**Comment by johnnyreilly** [comment]
> "I’m pretty lenient since I’m still an experiment, "

@andrewbranch - I'm very much enjoying the "tone" of the @typescript-bot 😁

**Comment by Tan90Qian** [comment]
@jgoz so what should I do? I just use typescript for about one month.

**Comment by Tan90Qian** [comment]
```typescript
// shallow/index.d.ts
render(element: ReactElement, context?: any): void | ReactElement;

// some test case
const component = renderer.render(
    <TableForm key="test" onSubmit={onSubmit} form={form} render={renderProps} />
);
expect(component).toMatchSnapshot();
if(component) {
    expect(component.type).toBe('div');
}
```
Is this ok?

File: types/react-test-renderer/shallow/index.d.ts [hunk 1]
```diff
@@ -14,9 +14,10 @@ export interface ShallowRenderer {
      */
     getRenderOutput(): ReactElement;
     /**
-     * Similar to `ReactDOM.render` but it doesn't require DOM and only renders a single level deep.
+     * Similar to `ReactDOM.render` but it doesn't require DOM and only renders a single level deep,
+     * and returns shallowly rendered output.
      */
-    render(element: ReactElement, context?: any): void;
+    render(element: ReactElement, context?: any): ReactElement;
```
[the remark this answers is not in the document]
Comment (Jessidhia) [reply types/react-test-renderer/shallow/index.d.ts]: If we had strongly typed `ReactElement`, we'd be able to restrict the input type to only be custom components, as well as make it return the same type as the function/class component returns. 😔

**Review by Jessidhia — COMMENTED** [review]
This also needs tests; an easy one could just be to add an `// $ExpectType null | string | number | ...` comment on the line before the `shallowRenderer.render(component);` call in `react-test-renderer-tests.ts`.

---

@jgoz: It seems the `this._rendering` check is just a guard against reentrancy -- you're not supposed to be able to call `renderer.render` from inside a component that is itself being rendered by the same `renderer.render`.

Given that it's something that can only happens when the API is being severely misused, as opposed to a possible regular return value, I think it's OK to not include it as a possible return value here.

File: types/react-test-renderer/shallow/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tan90Qian) [reply types/react-test-renderer/shallow/index.d.ts]: I think there won't return string or number. This render api is Similar to `ReactDOM.render`, so the `element` param must be a JSX element or a constructor function for JSX elements.

**Comment by jgoz** [comment]
@jessidhia 
> It seems the this._rendering check is just a guard against reentrancy -- you're not supposed to be able to call renderer.render from inside a component that is itself being rendered by the same renderer.render.

👍 good point.

File: types/react-test-renderer/shallow/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jgoz) [reply types/react-test-renderer/shallow/index.d.ts]: I agree with @jessidhia here that both `getRenderOutput()` and `render()` should return `ReactNode` (and that `ReactNode` is not correct).

I think the argument to `render()` can remain `ReactElement` (though according to the actual code, it also accepts `null`, but throws).

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]
Looks like the bot might have accidentally read @jgoz's comment as an approval.
I'll mark as changes requested until the comments are addressed.
